### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] using parameter as object computed key is not detected

### ISSUE THREAD

**Issue by abrahamguo** [typescript-eslint/typescript-eslint#9531]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=MYewdgzgLgBA1gUwJ4wLwwDwBUYIB5QJgAmEM0ATgJZgDmAfABQCGAXDFgJRr0yMDeMANrMAuuygUArghgBfTgG4AUKEiwAbswA2aTDnyESZSjQYt2XHn0GSZ7ZvKWrw0GMwoU92XASKlySTMmNg5uVF4RUUUgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=&tokens=false

### Repro Code

```TypeScript
const key = <T extends string>(a: T) => ({ [a]: true });
const val = <T extends string>(a: T) => ({ true: a });
const arr = <T extends string>(a: T) => [a];
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unused-type-parameters": "error",
  },
};
```

### Expected Result

The rule should not report on any of the three functions.

### Actual Result

The rule reports on the `key` function but not the other two.

**Comment by abrahamguo** [comment]
Oops, I think this is actually correct, because TypeScript doesn't even add the key to the type.
